enhancements/etcd/automated-defrag.md

```diff
@@ -0,0 +1,235 @@
+
+[openshift/cluster-etcd-operator PR #625](https://github.com/openshift/cluster-etcd-operator/pull/625)
+
+### User Stories
+
+- I want my cluster to manage maintenance of etcd backend without interaction or planning by admin.
+- I want to ensure that heavy churn workloads do not negatively impact cluster performance.
+- I want to ensure that cluster operands utilize only the necessary resources to provide service.
+
+Large production clusters such as OpenShift CI feel the pain of fragmentation daily. The constant churn of resources 
+from creating and deleting many jobs can quickly lead to resource bloat as the result of fragmentation. This growth can 
+lead to OOM events and downtime. This controller was prototyped as a result of a CI outage.
+
+### Implementation Details/Notes/Constraints
+
+## etcd Storage Backend
+etcd's storage backend is powered by `bbolt` a B+Tree memory mapped key/value store with MVCC support. While MVCC is
+necessary to support fully concurrent transactions and the heart of the `Watch` functionality.
+MVCC poses a performance challenge for `bbolt` because the db can store many versions of the keys value. These
+versions can build up indefinitely and result in memory and disk performance issues. The solution is to regularly
+compact the key space.
+
+compaction is essentially a delete action in `bbolt`, where the prevision versions of a keys value are pruned.
+`bbolt` organizes data on pages and when a page is no longer used as the result of a db mutation it is called a
+free page and an accounting of all free pages are stored in a freelist for later reuse.
+
+Compaction does reduce the memory footprint by removing old keys but free pages can not be reclaimed back to disk
+without defragmentation.
+
+## Defrag Controller
+
+During the operator sync the controller will make a  call to the cluster API using the `MemberListRequest` RPC which 
+returns the current list of etcd members. We will then loop these members and for each dial the maintenance API using
+the `StatusRequest` RPC which will return detailed information on the backend status.
+
+```go
+type StatusResponse struct {
+    // dbSize is the size of the backend database physically allocated, in bytes, of the responding member.
+    DbSize int64 `protobuf:"varint,3,opt,name=dbSize,proto3" json:"dbSize,omitempty"`
+    [...] 
+    // dbSizeInUse is the size of the backend database logically in use, in bytes, of the responding member.
+    DbSizeInUse int64 `protobuf:"varint,9,opt,name=dbSizeInUse,proto3" json:"dbSizeInUse,omitempty"`
+}
+```
+
+From these two fields we can conclude the total fragmented space on that etcd members backend store. If the
+percentage of bytes of the total keyspace are over a defined threshold percentage. The controller will 
+utilize maintenance API again and issue a `DefragmentRequest` RPC.
+
+Defragmentation is I/O blocking thus important to only defrag a single etcd at a time. Additionally, to reduce
```

Comment:
How is this going to be achieved, to restrain defrag to a single etcd at a time? Could we detail that a bit, sorry if I missed it somewhere?

Reply:
will add in more detail